### PR TITLE
Implement new rule no-restricted-paths (fixes #155)

### LINKED ISSUE #155: Rule proposal: no-restricted-paths
Given the following folder structure of a project:

```
my-project
├── client
│   └── foo.js
└── server
    └── bar.js
```

I want to disallow that any file in the `client` folder can import any module from the `server` folder.

The following should be invalid in `client/foo.js`:

```js
import bar from '../server/bar.js';
```

### PR DESCRIPTION
**Pull request by lo1tuma**

### CODE REVIEW

File: docs/rules/no-restricted-paths.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+The following patterns are considered problems when configuration set to `[ { "target": "./client", "from": "./server" } ]`:
+
+```js
+import bar from '../server/bar';
+```
+
+The following patterns are not considered problems when configuration set to `[ { "target": "./client", "from": "./server" } ]`:
+
+```js
+import foo from '../server/baz';
```
Comment (jfmengels) [inline docs/rules/no-restricted-paths.md]: I'm not sure I see how this is different, except for the fact that `server/baz` does not exist.

File: docs/rules/no-restricted-paths.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# no-restricted-paths - Restrict which files can be imported in a given folder
+
+Some projects contain files which are not always meant to be executed in the same environment.
+For example consider a web application that contains specific code for the server and some specific code for the browser/client. In this case you don’t want to import server-only files in your client code.
+
+In order to prevent such scenarios this rule allows you to define restricted zones where you can forbid files from imported if they match a specific path.
+
+## Rule Details
+
+This rule has tow options. The first option is an array to define all restricted zones. The second option is optional and allows you to override the `basePath` which is the current working directory by default. All paths defined in the first option are either absolute paths or relative to the `basePath`.
```
Comment (jfmengels) [inline docs/rules/no-restricted-paths.md]: typo: `tow` --> `two`.

File: tests/src/rules/no-restricted-paths.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      ],
+    }),
+    test({
+      code: 'const foo = require("../client/foo.js")',
+      filename: testFilePath('./restricted-paths/client/a.js'),
+      options: [
+        [ { target: './tests/files/restricted-paths/client', from: './tests/files/restricted-paths/server' } ],
+      ],
+    }),
+  ],
```
Comment (jfmengels) [inline tests/src/rules/no-restricted-paths.js]: Mind adding some tests where the filename is in the `server` folder, for completeness' sake?

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 
 ## [Unreleased]
 ### Added
+- New rule [`no-restricted-paths`]. ([#155], thanks [@lo1tuma])
```
Comment (jfmengels) [inline CHANGELOG.md]: `Add new rule ...`, and you can also add #371 now that you've opened a PR ;)

**Comment by jfmengels** [comment]
A few comments, but this looks really good, thanks @lo1tuma!

File: docs/rules/no-restricted-paths.md (hunk 1 above)
Comment (lo1tuma) [inline docs/rules/no-restricted-paths.md]: My fault, it should be `import baz from '../client/baz'`.

**Comment by lo1tuma** [comment]
@jfmengels Thanks for the review. All comments have been addressed.

**Comment by jfmengels** [comment]
Great work @lo1tuma, LGTM :)

**Comment by gilligan** [comment]
Any chance of merging this sometime soon?

**Pull request merged**